action/protocol/staking/read_state.go

```diff
@@ -54,21 +53,21 @@ func readStateBucketsByVoter(ctx context.Context, sr protocol.StateReader,
 	return toIoTeXTypesVoteBucketList(buckets)
 }
 
-func readStateBucketsByCandidate(ctx context.Context, sr protocol.StateReader, cc CandidateCenter,
+func readStateBucketsByCandidate(ctx context.Context, csr CandidateStateReader, cc CandidateCenter,
 	req *iotexapi.ReadStakingDataRequest_VoteBucketsByCandidate) (*iotextypes.VoteBucketList, error) {
 	c := cc.GetByName(req.GetCandName())
 	if c == nil {
 		return &iotextypes.VoteBucketList{}, nil
 	}
 
-	indices, err := getCandBucketIndices(sr, c.Owner)
+	indices, err := csr.getCandBucketIndices(c.Owner)
```

Comment:
the reason we need to create candidate center outside is that we need to interact with candidate center to get candidate by name. What if candidate center is part of csr?